// reflex/dist/reflex.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./task", "./signal", "./effects", "./dom"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./task"), require("./signal"), require("./effects"), require("./dom"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.task, global.signal, global.effects, global.dom);
    global.application = mod.exports;
  }
})(this, function (exports, _task, _signal, _effects, _dom) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.start = exports.beginner = undefined;

  var _slicedToArray = function () {
    function sliceIterator(arr, i) {
      var _arr = [];
      var _n = true;
      var _d = false;
      var _e = undefined;

      try {
        for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
          _arr.push(_s.value);

          if (i && _arr.length === i) break;
        }
      } catch (err) {
        _d = true;
        _e = err;
      } finally {
        try {
          if (!_n && _i["return"]) _i["return"]();
        } finally {
          if (_d) throw _e;
        }
      }

      return _arr;
    }

    return function (arr, i) {
      if (Array.isArray(arr)) {
        return arr;
      } else if (Symbol.iterator in Object(arr)) {
        return sliceIterator(arr, i);
      } else {
        throw new TypeError("Invalid attempt to destructure non-iterable instance");
      }
    };
  }();

  var first = function first(xs) {
    return xs[0];
  };
  var second = function second(xs) {
    return xs[1];
  };

  var beginner = exports.beginner = function beginner(configuration) {
    return { flags: void 0,
      init: function init(_) {
        return [configuration.model, _effects.Effects.none];
      },
      update: function update(model, action) {
        return [configuration.update(model, action), _effects.Effects.none];
      },
      view: configuration.view
    };
  };

  var start = exports.start = function start(configuration) {
    var init = configuration.init;
    var view = configuration.view;
    var update = configuration.update;
    var flags = configuration.flags;

    var _mailbox = (0, _signal.mailbox)({});

    var address = _mailbox.address;
    var signal = _mailbox.signal;

    var step = function step(_ref, action) {
      var _ref2 = _slicedToArray(_ref, 2);

      var model = _ref2[0];
      var _ = _ref2[1];
      return update(model, action);
    };

    var display = function display(model) {
      return (0, _dom.root)(view, model, address);
    };

    var steps = (0, _signal.reductions)(step, init(flags), signal);

    var model = (0, _signal.map)(first, steps);
    var application = { address: address,
      model: model,
      task: (0, _signal.map)(second, steps),
      view: (0, _signal.map)(display, model)
    };

    return application;
  };
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.dom = mod.exports;
  }
})(this, function (exports) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  function _toConsumableArray(arr) {
    if (Array.isArray(arr)) {
      for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i];

      return arr2;
    } else {
      return Array.from(arr);
    }
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  var driver = null;
  var absent = new String("absent");

  var VirtualRoot = function () {
    function VirtualRoot(view, model, address) {
      _classCallCheck(this, VirtualRoot);

      this.view = view;
      this.model = model;
      this.address = address;
    }

    _createClass(VirtualRoot, [{
      key: "renderWith",
      value: function renderWith(current) {
        var exception = absent;
        var previous = driver;
        driver = current;

        try {
          driver.render(this.view(this.model, this.address));
        } catch (error) {
          exception = error;
        }

        driver = previous;

        if (exception != absent) {
          throw exception;
        }
      }
    }]);

    return VirtualRoot;
  }();

  VirtualRoot.prototype.$type = "VirtualRoot";

  var LazyNode = exports.LazyNode = function () {
    function LazyNode(tagName, properties, children) {
      _classCallCheck(this, LazyNode);

      this.tagName = tagName;
      this.properties = properties;
      this.children = children;
      this.key = properties == null ? null : properties.key;
    }

    _createClass(LazyNode, [{
      key: "force",
      value: function force() {
        if (driver == null) {
          throw TypeError('LazyTree may only be forced from with in the Root.renderWith(driver) call');
        }

        return driver.node(this.tagName, this.properties, this.children);
      }
    }]);

    return LazyNode;
  }();

  LazyNode.prototype.$type = "LazyTree";

  var LazyThunk = function () {
    function LazyThunk(key, view, args) {
      _classCallCheck(this, LazyThunk);

      this.key = key;
      this.view = view;
      this.args = args;
    }

    _createClass(LazyThunk, [{
      key: "force",
      value: function force() {
        var _driver;

        if (driver == null) {
          throw TypeError('LazyTree may only be forced from with in the Root.renderWith(driver) call');
        }

        return (_driver = driver).thunk.apply(_driver, [this.key, this.view].concat(_toConsumableArray(this.args)));
      }
    }]);

    return LazyThunk;
  }();

  LazyThunk.prototype.$type = "LazyTree";

  var root = exports.root = function root(view, model, address) {
    return new VirtualRoot(view, model, address);
  };

  var text = exports.text = function text(content) {
    return driver == null ? content : driver.text == null ? content : driver.text(content);
  };

  var node = exports.node = function node(tagName, properties, children) {
    return driver == null ? new LazyNode(tagName, properties, children) : driver.node(tagName, properties, children);
  };

  var thunk = exports.thunk = function thunk(key, view) {
    for (var _len = arguments.length, args = Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
      args[_key - 2] = arguments[_key];
    }

    var _driver2;

    return driver == null ? new LazyThunk(key, view, args) : (_driver2 = driver).thunk.apply(_driver2, [key, view].concat(args));
  };
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', './task'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./task'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.task);
    global.effects = mod.exports;
  }
})(this, function (exports, _task2) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Effects = undefined;

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  var raise = function raise(error) {
    throw Error('Effects should be created from task that never fail but it did fail with error ' + error);
  };

  var ignore = function ignore(_) {
    return void 0;
  };

  var nil = _task2.Task.succeed(void 0);

  var never = new _task2.Task(function (succeed, fail) {
    return void 0;
  });

  var Effects = function () {
    function Effects() {
      _classCallCheck(this, Effects);
    }

    _createClass(Effects, null, [{
      key: 'task',
      value: function task(_task) {
        console.warn('Effects.task is deprecated please use Effects.perform instead');
        return new Perform(_task);
      }
    }, {
      key: 'perform',
      value: function perform(task) {
        return new Perform(task);
      }
    }, {
      key: 'tick',
      value: function tick(tag) {
        console.warn('Effects.tick is deprecated please use Effects.perform(Task.requestAnimationFrame().map(tag)) instead');
        return new Perform(_task2.Task.requestAnimationFrame().map(tag));
      }
    }, {
      key: 'receive',
      value: function receive(action) {
        var fx = new Perform(new _task2.Task(function (succeed, fail) {
          return void Promise.resolve(action).then(succeed, fail);
        }));
        return fx;
      }
    }, {
      key: 'batch',
      value: function batch(effects) {
        return new Batch(effects);
      }
    }, {
      key: 'driver',
      value: function driver(address) {
        return function (fx) {
          if (!(fx instanceof None)) {
            _task2.Task.fork(fx.send(address), ignore, raise);
          }
        };
      }
    }]);

    return Effects;
  }();

  exports.Effects = Effects;

  var Perform = function (_Effects) {
    _inherits(Perform, _Effects);

    function Perform(task) {
      _classCallCheck(this, Perform);

      var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Perform).call(this));

      _this.task = task;
      return _this;
    }

    _createClass(Perform, [{
      key: 'map',
      value: function map(f) {
        return new Perform(this.task.map(f));
      }
    }, {
      key: 'send',
      value: function send(address) {
        return this.task.chain(function (value) {
          return _task2.Task.send(address, value);
        });
      }
    }]);

    return Perform;
  }(Effects);

  var None = function (_Effects2) {
    _inherits(None, _Effects2);

    function None() {
      _classCallCheck(this, None);

      return _possibleConstructorReturn(this, Object.getPrototypeOf(None).apply(this, arguments));
    }

    _createClass(None, [{
      key: 'map',
      value: function map(f) {
        return Effects.none;
      }
    }, {
      key: 'send',
      value: function send(address) {
        return nil;
      }
    }]);

    return None;
  }(Effects);

  Effects.none = new None();

  var Batch = function (_Effects3) {
    _inherits(Batch, _Effects3);

    function Batch(effects) {
      _classCallCheck(this, Batch);

      var _this3 = _possibleConstructorReturn(this, Object.getPrototypeOf(Batch).call(this));

      _this3.effects = effects;
      return _this3;
    }

    _createClass(Batch, [{
      key: 'map',
      value: function map(f) {
        return new Batch(this.effects.map(function (effect) {
          return effect.map(f);
        }));
      }
    }, {
      key: 'send',
      value: function send(address) {
        var _this4 = this;

        return new _task2.Task(function (succeed, fail) {
          var effects = _this4.effects;

          var count = effects.length;
          var index = 0;
          while (index < count) {
            var effect = effects[index];
            if (!(effect instanceof None)) {
              _task2.Task.fork(effect.send(address), ignore, raise);
            }

            index = index + 1;
          }
          succeed(void 0);
        });
      }
    }]);

    return Batch;
  }(Effects);
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./dom"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./dom"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.dom);
    global.html = mod.exports;
  }
})(this, function (exports, _dom) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.html = undefined;
  var html = exports.html = ["a", "abbr", "address", "area", "article", "aside", "audio", "b", "base", "bdi", "bdo", "big", "blockquote", "body", "br", "button", "canvas", "caption", "cite", "code", "col", "colgroup", "data", "datalist", "dd", "del", "details", "dfn", "dialog", "div", "dl", "dt", "em", "embed", "fieldset", "figcaption", "figure", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6", "head", "header", "hr", "html", "i", "iframe", "img", "input", "ins", "kbd", "keygen", "label", "legend", "li", "link", "main", "map", "mark", "menu", "menuitem", "meta", "meter", "nav", "noscript", "object", "ol", "optgroup", "option", "output", "p", "param", "picture", "pre", "progress", "q", "rp", "rt", "ruby", "s", "samp", "script", "section", "select", "small", "source", "span", "strong", "style", "sub", "summary", "sup", "table", "tbody", "td", "textarea", "tfoot", "th", "thead", "time", "title", "tr", "track", "u", "ul", "var", "video", "wbr", "circle", "clipPath", "defs", "ellipse", "g", "line", "linearGradient", "mask", "path", "pattern", "polygon", "polyline", "radialGradient", "rect", "stop", "svg", "text", "tspan"].reduce(function (html, tagName) {
    var element = function element(properties, children) {
      return (0, _dom.node)(tagName, properties, children);
    };

    html[tagName] = element;
    return html;
  }, Object.create(null));
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./dom", "./html", "./signal", "./application", "./effects", "./task"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./dom"), require("./html"), require("./signal"), require("./application"), require("./effects"), require("./task"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.dom, global.html, global.signal, global.application, global.effects, global.task);
    global.index = mod.exports;
  }
})(this, function (exports, _dom, _html, _signal, _application, _effects, _task) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  Object.defineProperty(exports, "thunk", {
    enumerable: true,
    get: function () {
      return _dom.thunk;
    }
  });
  Object.defineProperty(exports, "node", {
    enumerable: true,
    get: function () {
      return _dom.node;
    }
  });
  Object.defineProperty(exports, "text", {
    enumerable: true,
    get: function () {
      return _dom.text;
    }
  });
  Object.defineProperty(exports, "html", {
    enumerable: true,
    get: function () {
      return _html.html;
    }
  });
  Object.defineProperty(exports, "forward", {
    enumerable: true,
    get: function () {
      return _signal.forward;
    }
  });
  Object.defineProperty(exports, "mailbox", {
    enumerable: true,
    get: function () {
      return _signal.mailbox;
    }
  });
  Object.defineProperty(exports, "start", {
    enumerable: true,
    get: function () {
      return _application.start;
    }
  });
  Object.defineProperty(exports, "beginner", {
    enumerable: true,
    get: function () {
      return _application.beginner;
    }
  });
  Object.defineProperty(exports, "Effects", {
    enumerable: true,
    get: function () {
      return _effects.Effects;
    }
  });
  Object.defineProperty(exports, "Task", {
    enumerable: true,
    get: function () {
      return _task.Task;
    }
  });
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.preemptiveAnimationFrame = mod.exports;
  }
})(this, function (exports) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var NO_REQUEST = 0;
  var PENDING_REQUEST = 1;
  var EXTRA_REQUEST = 2;

  var nextID = 0;
  var state = NO_REQUEST;
  var requests = [];
  var ids = [];

  var absent = new String("absent");

  var requestAnimationFrame = exports.requestAnimationFrame = function requestAnimationFrame(request) {
    if (state === NO_REQUEST) {
      window.requestAnimationFrame(performAnimationFrame);
    }

    var id = ++nextID;
    requests.push(request);
    ids.push(id);
    state = PENDING_REQUEST;
    return id;
  };

  var cancelAnimationFrame = exports.cancelAnimationFrame = function cancelAnimationFrame(id) {
    var index = ids.indexOf(id);
    if (index >= 0) {
      ids.splice(index, 1);
      requests.splice(index, 1);
    }
  };

  var forceAnimationFrame = exports.forceAnimationFrame = function forceAnimationFrame() {
    var time = arguments.length <= 0 || arguments[0] === undefined ? window.performance.now() : arguments[0];
    return performAnimationFrame(time);
  };

  var performAnimationFrame = function performAnimationFrame(time) {
    switch (state) {
      case NO_REQUEST:
        throw Error("Unexpected frame request");
      case PENDING_REQUEST:
        window.requestAnimationFrame(performAnimationFrame);
        state = EXTRA_REQUEST;
        ids.splice(0);
        dispatchAnimationFrame(requests.splice(0), 0, time);
        break;
      case EXTRA_REQUEST:
        state = NO_REQUEST;
        break;
    }
  };

  var dispatchAnimationFrame = function dispatchAnimationFrame(requests, index, time) {
    var exception = absent;
    var count = requests.length;
    try {
      while (index < count) {
        var request = requests[index];
        index = index + 1;
        request(time);
      }
    } catch (error) {
      exception = error;
    }

    if (index < count) {
      dispatchAnimationFrame(requests, index, time);
    }

    if (exception != absent) {
      throw exception;
    }
  };
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.signal = mod.exports;
  }
})(this, function (exports) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  var absent = new String("absent");

  var Input = function () {
    _createClass(Input, null, [{
      key: "Address",
      value: function Address(signal) {
        if (signal.address == null) {
          signal.address = signal.receive.bind(signal);

          return signal.address;
        } else {
          return signal.address;
        }
      }
    }, {
      key: "notify",
      value: function notify(message, addressBook, from, to) {
        var exception = absent;
        try {
          while (from < to) {
            var address = addressBook[from];
            if (address != null) {
              address(message);
            }
            from = from + 1;
          }
        } catch (error) {
          exception = error;
        }

        if (from < to) {
          Input.notify(message, addressBook, from + 1, to);
        }

        if (exception != absent) {
          throw exception;
        }
      }
    }, {
      key: "connect",
      value: function connect(signal, address) {
        if (signal.addressBook == null) {
          signal.addressBook = [address];
        } else {
          var addressBook = signal.addressBook;
          if (addressBook.indexOf(address) < 0) {
            addressBook.push(address);
          }
        }
      }
    }]);

    function Input(value) {
      _classCallCheck(this, Input);

      this.value = value;
      this.isBlocked = false;

      this.addressBook = null;
      this.queue = null;
      this.address = null;
    }

    _createClass(Input, [{
      key: "receive",
      value: function receive(value) {
        if (this.isBlocked) {
          if (this.queue == null) {
            this.queue = [value];
          } else {
            this.queue.push(value);
          }
        } else {
          var exception = absent;
          this.isBlocked = true;
          try {
            this.value = value;

            if (this.addressBook != null) {
              var addressBook = this.addressBook;
              Input.notify(value, addressBook, 0, addressBook.length);
            }
          } catch (error) {
            exception = error;
          }

          this.isBlocked = false;
          if (this.queue != null && this.queue.length > 0) {
            this.receive(value = this.queue.shift());
          }

          if (exception != absent) {
            throw exception;
          }
        }
      }
    }, {
      key: "subscribe",
      value: function subscribe(address) {
        Input.connect(this, address);
        address(this.value);
      }
    }, {
      key: "connect",
      value: function connect(address) {
        if (this.addressBook == null) {
          this.addressBook = [address];
        } else {
          var addressBook = this.addressBook;
          if (addressBook.indexOf(address) < 0) {
            addressBook.push(address);
          }
        }
      }
    }]);

    return Input;
  }();

  Input.prototype.$type = "Signal.Signal";

  var Inbox = function Inbox(message) {
    _classCallCheck(this, Inbox);

    this.signal = new Input(message);
    this.address = Input.Address(this.signal);
  };

  Inbox.prototype.$type = "Signal.Mailbox";

  var mailbox = exports.mailbox = function mailbox(message) {
    return new Inbox(message);
  };

  var Forward = function Forward(address, tag) {
    var forward = function forward(message) {
      return address(tag(message));
    };
    forward.to = address;
    forward.tag = tag;
    return forward;
  };

  if (global['reflex/address'] == null) {
    global['reflex/address'] = 0;
  }

  var forward = exports.forward = function forward(address, tag) {
    var id = address.id != null ? address.id : address.id = global['reflex/address']++;
    var key = "reflex/address/" + id;

    return tag[key] || (tag[key] = Forward(address, tag));
  };

  var reductions = exports.reductions = function reductions(step, state, input) {
    var output = new Input(state);
    input.connect(forward(Input.Address(output), function (value) {
      return step(output.value, value);
    }));
    return output;
  };

  var map = exports.map = function map(f, input) {
    var output = new Input(f(input.value));
    input.connect(forward(Input.Address(output), f));
    return output;
  };
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', './preemptive-animation-frame'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./preemptive-animation-frame'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.preemptiveAnimationFrame);
    global.task = mod.exports;
  }
})(this, function (exports, _preemptiveAnimationFrame) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Task = undefined;

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  var Task = exports.Task = function () {
    _createClass(Task, null, [{
      key: 'create',
      value: function create(execute) {
        console.warn('Task.create is deprecated API use new Task instead');
        return new Task(execute);
      }
    }, {
      key: 'future',
      value: function future(request) {
        console.warn('Task.future is deprecated API use new Task instead');
        return new Future(request);
      }
    }, {
      key: 'succeed',
      value: function succeed(value) {
        return new Succeed(value);
      }
    }, {
      key: 'fail',
      value: function fail(error) {
        return new Fail(error);
      }
    }, {
      key: 'spawn',
      value: function spawn(task) {
        return new Spawn(task);
      }
    }, {
      key: 'sleep',
      value: function sleep(time) {
        return new Sleep(time);
      }
    }, {
      key: 'requestAnimationFrame',
      value: function requestAnimationFrame() {
        return new AnimationFrame();
      }
    }, {
      key: 'send',
      value: function send(address, message) {
        return new Send(address, message);
      }
    }, {
      key: 'fork',
      value: function fork(task, onSucceed, onFail) {
        return Process.fork(task, onSucceed, onFail);
      }
    }]);

    function Task(execute, cancel) {
      _classCallCheck(this, Task);

      if (execute != null) {
        this.fork = execute;
      }
      if (cancel != null) {
        this.abort = cancel;
      }
    }

    _createClass(Task, [{
      key: 'chain',
      value: function chain(next) {
        return new Chain(this, next);
      }
    }, {
      key: 'map',
      value: function map(f) {
        return new Map(this, f);
      }
    }, {
      key: 'capture',
      value: function capture(handle) {
        return new Capture(this, handle);
      }
    }, {
      key: 'format',
      value: function format(f) {
        return new Format(this, f);
      }
    }, {
      key: 'recover',
      value: function recover(regain) {
        return new Recover(this, regain);
      }
    }, {
      key: 'fork',
      value: function fork(succeed, fail) {}
    }, {
      key: 'abort',
      value: function abort(handle) {}
    }]);

    return Task;
  }();

  var Succeed = function (_Task) {
    _inherits(Succeed, _Task);

    function Succeed(value) {
      _classCallCheck(this, Succeed);

      var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Succeed).call(this));

      _this.value = value;
      return _this;
    }

    _createClass(Succeed, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        succeed(this.value);
      }
    }]);

    return Succeed;
  }(Task);

  var Fail = function (_Task2) {
    _inherits(Fail, _Task2);

    function Fail(error) {
      _classCallCheck(this, Fail);

      var _this2 = _possibleConstructorReturn(this, Object.getPrototypeOf(Fail).call(this));

      _this2.error = error;
      return _this2;
    }

    _createClass(Fail, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        fail(this.error);
      }
    }]);

    return Fail;
  }(Task);

  var Sleep = function (_Task3) {
    _inherits(Sleep, _Task3);

    function Sleep(time) {
      _classCallCheck(this, Sleep);

      var _this3 = _possibleConstructorReturn(this, Object.getPrototypeOf(Sleep).call(this));

      _this3.time = time;
      return _this3;
    }

    _createClass(Sleep, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        return setTimeout(succeed, this.time, void 0);
      }
    }, {
      key: 'abort',
      value: function abort(id) {
        clearTimeout(id);
      }
    }]);

    return Sleep;
  }(Task);

  var AnimationFrame = function (_Task4) {
    _inherits(AnimationFrame, _Task4);

    function AnimationFrame() {
      _classCallCheck(this, AnimationFrame);

      return _possibleConstructorReturn(this, Object.getPrototypeOf(AnimationFrame).call(this));
    }

    _createClass(AnimationFrame, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        return (0, _preemptiveAnimationFrame.requestAnimationFrame)(succeed);
      }
    }, {
      key: 'abort',
      value: function abort(id) {
        (0, _preemptiveAnimationFrame.cancelAnimationFrame)(id);
      }
    }]);

    return AnimationFrame;
  }(Task);

  var threadID = 0;

  var Spawn = function (_Task5) {
    _inherits(Spawn, _Task5);

    function Spawn(task) {
      _classCallCheck(this, Spawn);

      var _this5 = _possibleConstructorReturn(this, Object.getPrototypeOf(Spawn).call(this));

      _this5.task = task;
      return _this5;
    }

    _createClass(Spawn, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        var _this6 = this;

        Promise.resolve(null).then(function (_) {
          return Task.fork(_this6.task, noop, noop);
        });

        succeed(++threadID);
      }
    }]);

    return Spawn;
  }(Task);

  var Send = function (_Task6) {
    _inherits(Send, _Task6);

    function Send(address, message) {
      _classCallCheck(this, Send);

      var _this7 = _possibleConstructorReturn(this, Object.getPrototypeOf(Send).call(this));

      _this7.message = message;
      _this7.address = address;
      return _this7;
    }

    _createClass(Send, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        succeed(void this.address(this.message));
      }
    }]);

    return Send;
  }(Task);

  var Future = function (_Task7) {
    _inherits(Future, _Task7);

    function Future(request) {
      _classCallCheck(this, Future);

      var _this8 = _possibleConstructorReturn(this, Object.getPrototypeOf(Future).call(this));

      _this8.request = request;
      return _this8;
    }

    _createClass(Future, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        this.request().then(succeed, fail);
      }
    }]);

    return Future;
  }(Task);

  var Then = function (_Task8) {
    _inherits(Then, _Task8);

    function Then(task) {
      _classCallCheck(this, Then);

      var _this9 = _possibleConstructorReturn(this, Object.getPrototypeOf(Then).call(this));

      _this9.task = task;
      return _this9;
    }

    _createClass(Then, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        var _this10 = this;

        this.task.fork(function (value) {
          return void _this10.next(value).fork(succeed, fail);
        }, fail);
      }
    }]);

    return Then;
  }(Task);

  var Chain = function (_Then) {
    _inherits(Chain, _Then);

    function Chain(task, next) {
      _classCallCheck(this, Chain);

      var _this11 = _possibleConstructorReturn(this, Object.getPrototypeOf(Chain).call(this, task));

      _this11.next = next;
      return _this11;
    }

    return Chain;
  }(Then);

  var Map = function (_Then2) {
    _inherits(Map, _Then2);

    function Map(task, mapper) {
      _classCallCheck(this, Map);

      var _this12 = _possibleConstructorReturn(this, Object.getPrototypeOf(Map).call(this, task));

      _this12.mapper = mapper;
      return _this12;
    }

    _createClass(Map, [{
      key: 'next',
      value: function next(input) {
        return new Succeed(this.mapper(input));
      }
    }]);

    return Map;
  }(Then);

  var Catch = function (_Task9) {
    _inherits(Catch, _Task9);

    function Catch(task) {
      _classCallCheck(this, Catch);

      var _this13 = _possibleConstructorReturn(this, Object.getPrototypeOf(Catch).call(this));

      _this13.task = task;
      return _this13;
    }

    _createClass(Catch, [{
      key: 'fork',
      value: function fork(succeed, fail) {
        var _this14 = this;

        this.task.fork(succeed, function (error) {
          return void _this14.handle(error).fork(succeed, fail);
        });
      }
    }]);

    return Catch;
  }(Task);

  var Capture = function (_Catch) {
    _inherits(Capture, _Catch);

    function Capture(task, handle) {
      _classCallCheck(this, Capture);

      var _this15 = _possibleConstructorReturn(this, Object.getPrototypeOf(Capture).call(this, task));

      _this15.handle = handle;
      return _this15;
    }

    return Capture;
  }(Catch);

  var Recover = function (_Catch2) {
    _inherits(Recover, _Catch2);

    function Recover(task, regain) {
      _classCallCheck(this, Recover);

      var _this16 = _possibleConstructorReturn(this, Object.getPrototypeOf(Recover).call(this, task));

      _this16.regain = regain;
      return _this16;
    }

    _createClass(Recover, [{
      key: 'handle',
      value: function handle(error) {
        return new Succeed(this.regain(error));
      }
    }]);

    return Recover;
  }(Catch);

  var Format = function (_Catch3) {
    _inherits(Format, _Catch3);

    function Format(task, formatter) {
      _classCallCheck(this, Format);

      var _this17 = _possibleConstructorReturn(this, Object.getPrototypeOf(Format).call(this, task));

      _this17.formatter = formatter;
      return _this17;
    }

    _createClass(Format, [{
      key: 'handle',
      value: function handle(error) {
        return new Fail(this.formatter(error));
      }
    }]);

    return Format;
  }(Catch);

  var noop = function noop() {
    return void 0;
  };

  var nextID = 0;

  var Process = function () {
    _createClass(Process, null, [{
      key: 'fork',
      value: function fork(task, onSucceed, onFail) {
        var process = new Process(task);
        process.succeed = onSucceed;
        process.fail = onFail;
        process.schedule();
        return process;
      }
    }]);

    function Process(task) {
      _classCallCheck(this, Process);

      this.id = ++nextID;
      this.position = 0;
      this.root = task;
      this.stack = [];
      this.mailbox = [];
      this.abortHandle = null;
      this.isActive = true;
      this.isPending = false;
      this.success = null;
      this.failure = null;
      this.succeed = null;
      this.fail = null;
      this.onSucceed = this.onSucceed.bind(this);
      this.onFail = this.onFail.bind(this);
    }

    _createClass(Process, [{
      key: 'onSucceed',
      value: function onSucceed(value) {
        if (this.isPending) {
          this.isPending = false;
          this.abortHandle = null;

          if (this.success != null) {
            this.success.value = value;
          } else {
            this.success = new Succeed(value);
          }

          this.root = this.success;
          this.schedule();
        }
      }
    }, {
      key: 'onFail',
      value: function onFail(error) {
        if (this.isPending) {
          this.isPending = false;
          this.abortHandle = null;

          if (this.failure != null) {
            this.failure.error = error;
          } else {
            this.failure = new Fail(error);
          }

          this.root = this.failure;
          this.schedule();
        }
      }
    }, {
      key: 'kill',
      value: function kill(reason) {
        if (this.isActive) {
          this.isActive = false;
          if (this.root.abort) {
            this.root.abort(this.abortHandle);
          }
        }
      }
    }, {
      key: 'schedule',
      value: function schedule() {
        this.step();
      }
    }, {
      key: 'step',
      value: function step() {
        var process = this;
        while (process.isActive) {
          var task = process.root;
          if (task instanceof Succeed) {
            while (process.position < process.stack.length && process.stack[process.position] instanceof Catch) {
              process.position++;
            }

            if (process.position >= process.stack.length) {
              if (process.succeed != null) {
                process.succeed(task.value);
              }
              break;
            }

            var then = process.stack[process.position++];

            process.root = then.next(task.value);
            continue;
          }

          if (task instanceof Fail) {
            while (process.position < process.stack.length && process.stack[process.position] instanceof Then) {
              process.position++;
            }

            if (this.position >= process.stack.length) {
              if (process.fail != null) {
                process.fail(task.error);
              }
              break;
            }

            var _catch = process.stack[process.position++];

            process.root = _catch.handle(task.error);
            continue;
          }

          if (task instanceof Then) {
            if (process.position === 0) {
              process.stack.unshift(task);
            } else {
              process.stack[--process.position] = task;
            }

            process.root = task.task;

            continue;
          }

          if (task instanceof Catch) {
            if (process.position === 0) {
              process.stack.unshift(task);
            } else {
              process.stack[--process.position] = task;
            }

            process.root = task.task;

            continue;
          }

          if (task instanceof Task) {
            process.isPending = true;
            process.abortHandle = task.fork(process.onSucceed, process.onFail);
            break;
          }
        }
      }
    }]);

    return Process;
  }();
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["./test", "../"], factory);
  } else if (typeof exports !== "undefined") {
    factory(require("./test"), require("../"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.test, global._);
    global.testApi = mod.exports;
  }
})(this, function (_test, _) {
  "use strict";

  var _test2 = _interopRequireDefault(_test);

  var Reflex = _interopRequireWildcard(_);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) {
    return typeof obj;
  } : function (obj) {
    return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj;
  };

  (0, _test2.default)("test exported api", function (assert) {
    assert.ok(_typeof(Reflex.node), "function");

    assert.ok(_typeof(Reflex.html), "object");
    assert.ok(_typeof(Reflex.html.div), "function");

    assert.ok(_typeof(Reflex.thunk), "function");
    assert.ok(_typeof(Reflex.send), "function");
    assert.ok(_typeof(Reflex.forward), "function");

    assert.ok(_typeof(Reflex.Application), "function");

    assert.ok(_typeof(Reflex.Task.succeed), "function");
    assert.ok(_typeof(Reflex.Task.fail), "function");
    assert.ok(_typeof(Reflex.Task.io), "function");
    assert.ok(_typeof(Reflex.Task.onSuccess), "function");
    assert.ok(_typeof(Reflex.Task.onFailure), "function");
    assert.ok(_typeof(Reflex.Task.perform), "function");
    assert.ok(_typeof(Reflex.Task.run), "function");
  });
});
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "tape"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("tape"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.tape);
    global.test = mod.exports;
  }
})(this, function (exports, _tape) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var tape = _interopRequireWildcard(_tape);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  exports.default = function (description, unit) {
    return tape.test(description, function (test) {
      var result = unit(test);
      if (result && result.then) {
        result.then(function (_) {
          return test.end();
        }, function (error) {
          return test.end(error || true);
        });
      } else {
        test.end();
      }
    });
  };
});
