// reflex/lib/dom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var driver = null;
var absent = new String("absent");

var VirtualRoot = function () {
  function VirtualRoot(view, model, address) {
    _classCallCheck(this, VirtualRoot);

    this.view = view;
    this.model = model;
    this.address = address;
  }

  _createClass(VirtualRoot, [{
    key: "renderWith",
    value: function renderWith(current) {
      var exception = absent;
      var previous = driver;
      driver = current;

      try {
        driver.render(this.view(this.model, this.address));
      } catch (error) {
        exception = error;
      }

      driver = previous;

      if (exception != absent) {
        throw exception;
      }
    }
  }]);

  return VirtualRoot;
}();

VirtualRoot.prototype.$type = "VirtualRoot";

var LazyNode = exports.LazyNode = function () {
  function LazyNode(tagName, properties, children) {
    _classCallCheck(this, LazyNode);

    this.tagName = tagName;
    this.properties = properties;
    this.children = children;
    this.key = properties == null ? null : properties.key;
  }

  _createClass(LazyNode, [{
    key: "force",
    value: function force() {
      if (driver == null) {
        throw TypeError('LazyTree may only be forced from with in the Root.renderWith(driver) call');
      }

      return driver.node(this.tagName, this.properties, this.children);
    }
  }]);

  return LazyNode;
}();

LazyNode.prototype.$type = "LazyTree";

var LazyThunk = function () {
  function LazyThunk(key, view, args) {
    _classCallCheck(this, LazyThunk);

    this.key = key;
    this.view = view;
    this.args = args;
  }

  _createClass(LazyThunk, [{
    key: "force",
    value: function force() {
      var _driver;

      if (driver == null) {
        throw TypeError('LazyTree may only be forced from with in the Root.renderWith(driver) call');
      }

      return (_driver = driver).thunk.apply(_driver, [this.key, this.view].concat(_toConsumableArray(this.args)));
    }
  }]);

  return LazyThunk;
}();

LazyThunk.prototype.$type = "LazyTree";

var root = exports.root = function root(view, model, address) {
  return new VirtualRoot(view, model, address);
};

var text = exports.text = function text(content) {
  return driver == null ? content : driver.text == null ? content : driver.text(content);
};

var node = exports.node = function node(tagName, properties, children) {
  return driver == null ? new LazyNode(tagName, properties, children) : driver.node(tagName, properties, children);
};

var thunk = exports.thunk = function thunk(key, view) {
  for (var _len = arguments.length, args = Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
    args[_key - 2] = arguments[_key];
  }

  var _driver2;

  return driver == null ? new LazyThunk(key, view, args) : (_driver2 = driver).thunk.apply(_driver2, [key, view].concat(args));
};
