// reflex/lib/effects.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Effects = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _task2 = require('./task');

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var raise = function raise(error) {
  throw Error('Effects should be created from task that never fail but it did fail with error ' + error);
};

var ignore = function ignore(_) {
  return void 0;
};

var nil = _task2.Task.succeed(void 0);

var never = new _task2.Task(function (succeed, fail) {
  return void 0;
});

var Effects = function () {
  function Effects() {
    _classCallCheck(this, Effects);
  }

  _createClass(Effects, null, [{
    key: 'task',
    value: function task(_task) {
      console.warn('Effects.task is deprecated please use Effects.perform instead');
      return new Perform(_task);
    }
  }, {
    key: 'perform',
    value: function perform(task) {
      return new Perform(task);
    }
  }, {
    key: 'tick',
    value: function tick(tag) {
      console.warn('Effects.tick is deprecated please use Effects.perform(Task.requestAnimationFrame().map(tag)) instead');
      return new Perform(_task2.Task.requestAnimationFrame().map(tag));
    }
  }, {
    key: 'receive',
    value: function receive(action) {
      var fx = new Perform(new _task2.Task(function (succeed, fail) {
        return void Promise.resolve(action).then(succeed, fail);
      }));
      return fx;
    }
  }, {
    key: 'batch',
    value: function batch(effects) {
      return new Batch(effects);
    }
  }, {
    key: 'driver',
    value: function driver(address) {
      return function (fx) {
        if (!(fx instanceof None)) {
          _task2.Task.fork(fx.send(address), ignore, raise);
        }
      };
    }
  }]);

  return Effects;
}();

exports.Effects = Effects;

var Perform = function (_Effects) {
  _inherits(Perform, _Effects);

  function Perform(task) {
    _classCallCheck(this, Perform);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Perform).call(this));

    _this.task = task;
    return _this;
  }

  _createClass(Perform, [{
    key: 'map',
    value: function map(f) {
      return new Perform(this.task.map(f));
    }
  }, {
    key: 'send',
    value: function send(address) {
      return this.task.chain(function (value) {
        return _task2.Task.send(address, value);
      });
    }
  }]);

  return Perform;
}(Effects);

var None = function (_Effects2) {
  _inherits(None, _Effects2);

  function None() {
    _classCallCheck(this, None);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(None).apply(this, arguments));
  }

  _createClass(None, [{
    key: 'map',
    value: function map(f) {
      return Effects.none;
    }
  }, {
    key: 'send',
    value: function send(address) {
      return nil;
    }
  }]);

  return None;
}(Effects);

Effects.none = new None();

var Batch = function (_Effects3) {
  _inherits(Batch, _Effects3);

  function Batch(effects) {
    _classCallCheck(this, Batch);

    var _this3 = _possibleConstructorReturn(this, Object.getPrototypeOf(Batch).call(this));

    _this3.effects = effects;
    return _this3;
  }

  _createClass(Batch, [{
    key: 'map',
    value: function map(f) {
      return new Batch(this.effects.map(function (effect) {
        return effect.map(f);
      }));
    }
  }, {
    key: 'send',
    value: function send(address) {
      var _this4 = this;

      return new _task2.Task(function (succeed, fail) {
        var effects = _this4.effects;

        var count = effects.length;
        var index = 0;
        while (index < count) {
          var effect = effects[index];
          if (!(effect instanceof None)) {
            _task2.Task.fork(effect.send(address), ignore, raise);
          }

          index = index + 1;
        }
        succeed(void 0);
      });
    }
  }]);

  return Batch;
}(Effects);
